sentry/src/main/java/io/sentry/SentryOptions.java

```diff
@@ -1,60 +1,311 @@
+     *
+     * @param key the name of the configuration key
+     * @return the value of the configuration key or null if not found
      */
-    public SentryClientFactory getSentryClientFactory() {
-        return sentryClientFactory;
+    @Nullable
+    public String getConfigurationKey(String key) {
+        return lookup.get(key, dsn);
     }
 
     /**
-     * Sets the {@link SentryClientFactory} to be used when initializing the SDK.
-     * @param sentryClientFactory Factory used to create a {@link SentryClient}.
+     * Gets the lookup instance used by this options instance.
+     *
+     * @return the configured lookup instance
      */
-    public void setSentryClientFactory(SentryClientFactory sentryClientFactory) {
-        this.sentryClientFactory = sentryClientFactory;
+    public Lookup getLookup() {
+        return lookup;
     }
 
     /**
-     * Gets the DSN. If not DSN was set, the SDK will attempt to find one in the environment.
-     * @return Data Source Name.
+     * Gets the DSN instance used by this options instance.
+     *
+     * @return the configured DSN instance
      */
-    public String getDsn() {
+    public Dsn getDsn() {
         return dsn;
     }
 
     /**
-     * Sets the DSN to be used by the {@link SentryClient}.
-     * @param dsn Sentry Data Source Name.
+     * Gets the Sentry client factory instance used by this options instance.
+     * @return the configured client factory instance
      */
-    public void setDsn(String dsn) {
-        this.dsn = dsn;
+    public SentryClientFactory getClientFactory() {
+        return clientFactory;
     }
 
     /**
-     * Gets the {@link ResourceLoader} to be used when looking for properties.
-     * @return {@link ResourceLoader}
+     * Returns a new Sentry client obtained from the {@link #getClientFactory() factory}.
+     *
+     * @return the new sentry client or null if the client factory is invalid
      */
-    public ResourceLoader getResourceLoader() {
-        return resourceLoader;
+    @Nullable
+    public SentryClient createClient() {
+        return getClientFactory().createSentryClient(getDsn());
+    }
+
+    private static Dsn resolveDsn(Lookup lookup, @Nullable String dsn) {
+        try {
+            if (Util.isNullOrEmpty(dsn)) {
+                dsn = Dsn.dsnFrom(lookup);
+            }
+
+            return new Dsn(dsn);
+        } catch (Exception e) {
+            logger.error("Error creating valid DSN from: '{}'.", dsn, e);
+            throw e;
+        }
+    }
+
+    // the below is support for instantiating the default lookup
+
+    private static List<ConfigurationResourceLocator> getDefaultResourceLocators() {
+        return asList(new SystemPropertiesBasedLocator(), new EnvironmentBasedLocator(), new StaticFileLocator());
+    }
+
+    private static List<ConfigurationProvider> getDefaultHighPriorityConfigurationProviders() {
```

Comment:
The comment above regards POJO applies to these methods too. Would they fit better in some helper?